### PR TITLE
bpo-9453: fix SAX2DOM handling of processing instruction before first XML element

### PR DESCRIPTION
**Pull request by ghost**
https://bugs.python.org/issue9453

### CODE REVIEW

File: Lib/test/test_pulldom.py [hunk 1]
```diff
@@ -181,7 +182,6 @@ def test_thorough_parse(self):
         """Test some of the hard-to-reach parts of PullDOM."""
         self._test_thorough(pulldom.parse(None, parser=SAXExerciser()))
 
-    @unittest.expectedFailure
     def test_sax2dom_fail(self):
         """SAX2DOM can"t handle a PI before the root element."""
```
Comment (taleinat) [inline Lib/test/test_pulldom.py]: This doc-string should certainly be updated as part of this change.

File: Lib/test/test_pulldom.py [hunk 2]
```diff
@@ -280,6 +280,7 @@ class SAX2DOMExerciser(SAXExerciser):
     def parse(self, _):
         h = self._handler
         h.startDocument()
+        h.processingInstruction("target", "data")
```
Comment (taleinat) [inline Lib/test/test_pulldom.py:284]: This change should be reflected in SAX2DOMExerciser's doc-string.

File: Lib/test/test_pulldom.py [hunk 3]
```diff
@@ -205,10 +205,10 @@ def _test_thorough(self, pd, before_root=True):
             evt, node = next(pd)
             self.assertEqual(pulldom.COMMENT, evt)
             self.assertEqual("a comment", node.data)
-            evt, node = next(pd)
-            self.assertEqual(pulldom.PROCESSING_INSTRUCTION, evt)
-            self.assertEqual("target", node.target)
-            self.assertEqual("data", node.data)
+        evt, node = next(pd)
```
Comment (taleinat) [inline Lib/test/test_pulldom.py:209]: Nitpick: Add a newline before these? (just for consistency)

File: Lib/xml/dom/pulldom.py [hunk 4]
```diff
@@ -10,11 +10,12 @@
 IGNORABLE_WHITESPACE = "IGNORABLE_WHITESPACE"
 CHARACTERS = "CHARACTERS"
 
+
 class PullDOM(xml.sax.ContentHandler):
     _locator = None
     document = None
 
-    def __init__(self, documentFactory=None):
+    def __init__(self, documentFactory=None, handle_sax=False):
```
Comment (taleinat) [inline Lib/xml/dom/pulldom.py]: Does `handle_sax` need to be part of the public API like this? If this is just to allow the SAX2DOM sub-class to signal for different behavior in `buildDocument()`, perhaps just use a class attribute? If so, note that the `__init__` of the sub-class may be removed.

Also, please add a comment explaining the logical meaning of the `handle_sax` flag.

File: Lib/xml/dom/pulldom.py [hunk 5]
```diff
@@ -293,41 +298,46 @@ def clear(self):
...
         parentNode = self.elementStack[-2]
         parentNode.appendChild(curNode)
 
     def processingInstruction(self, target, data):
-        PullDOM.processingInstruction(self, target, data)
-        node = self.lastEvent[0][1]
-        parentNode = self.elementStack[-1]
-        parentNode.appendChild(node)
+        super().processingInstruction(target, data)
+        if self.lastEvent[0] is not None:
```
Comment (taleinat) [inline Lib/xml/dom/pulldom.py:322]: Was this a bug that is now fixed? Is there a test that exercises this case?

**Review by taleinat — CHANGES_REQUESTED** [review]
Looking good!

I only had a few rather minor comments.

**Comment by taleinat** [comment]
@jfgossage, note that this will require a NEWS entry. You should use [blurb](https://pypi.org/project/blurb/) to make one.

**Comment by taleinat** [comment]
@jfgossage, do you think you'll have a chance to continue your work on this PR soon?

File: Lib/test/test_pulldom.py (hunk 1 above)
Comment (ghost) [reply Lib/test/test_pulldom.py]: Agreed.

File: Lib/test/test_pulldom.py (hunk 2 above)
Comment (ghost) [reply Lib/test/test_pulldom.py:284]: Agreed.

File: Lib/test/test_pulldom.py (hunk 3 above)
Comment (ghost) [reply Lib/test/test_pulldom.py:209]: Agreed.

File: Lib/xml/dom/pulldom.py (hunk 5 above)
Comment (ghost) [reply Lib/xml/dom/pulldom.py:322]: This is the core of the fix for this issue. A test case already exists and removing the expected failure flag means that the PI test before root element will now be reported as successful when it succeeds. No new test case is needed.

File: Lib/xml/dom/pulldom.py (hunk 4 above)
Comment (ghost) [reply Lib/xml/dom/pulldom.py]: Handle-sax is used to make pullDOM support SAX2DOM specific processing, specifically creating a DOM tree while parsing. The only place it is used is for the PI handling that is being fixed.

I could change PullDOM to verify at runtime whether SAX2DOM is a subclass and thus keep everything within the implementation. Your choice and when you decide I will update the PR with all your suggestions.

Comment (taleinat) [reply Lib/xml/dom/pulldom.py]: I suggest making it a "protected" class attribute (start its name with an underscore), e.g. `_is_sax`. Have the parent class set it to `False` and `SAX2DOM` override it with `True`. `pullDOM` would indeed check it at run-time.

This would explicitly mark that we don't expect users to need to deal with it, perhaps only anyone writing sub-classes.

File: Lib/xml/dom/pulldom.py [hunk 6]
```diff
@@ -172,14 +174,17 @@ def buildDocument(self, uri, tagname):
         # Put everything we have seen so far into the document
         for e in self.pending_events:
             if e[0][0] == PROCESSING_INSTRUCTION:
-                _,target,data = e[0]
-                n = self.document.createProcessingInstruction(target, data)
+                _, target, data = e[0]
+                n = self.document.createProcessingInstruction(target,
```
Comment (taleinat) [inline Lib/xml/dom/pulldom.py:176]: This line split is unnecessary.

File: Lib/xml/dom/pulldom.py (hunk 5 above)
Comment (taleinat) [reply Lib/xml/dom/pulldom.py:322]: Ah, that makes sense, thanks for the clarification.

**Comment by jfgossage** [comment]
I have made the requested changes; please review again.

File: Lib/xml/dom/pulldom.py (hunk 6 above)
Comment (ghost) [reply Lib/xml/dom/pulldom.py:176]: `autopep8` complains if I don't split it.

Comment (taleinat) [reply Lib/xml/dom/pulldom.py:176]: Why?? The line's length is only 75 characters without splitting.

File: Lib/test/test_pulldom.py [hunk 7]
```diff
@@ -118,14 +118,15 @@ def test_expandItem(self):
             if evt == pulldom.START_ELEMENT:
                 break
         self.assertEqual("hr", node.tagName,
-            "expandNode did not leave DOMEventStream in the correct state.")
+                         "expandNode did not leave DOMEventStream"
```
Comment (taleinat) [inline Lib/test/test_pulldom.py:121]: As below, this line split is also unnecessary. It's best to making such changes.

(I guess autopep8 is being over-zealous.)

**Review by taleinat — CHANGES_REQUESTED** [review]
@jfgossage, please fix the two unnecessarily split lines.

File: Lib/xml/dom/pulldom.py [hunk 8]
```diff
@@ -293,41 +300,46 @@ def clear(self):
         self.parser = None
         self.stream = None
 
+
 class SAX2DOM(PullDOM):
+    def __init__(self, documentFactory=None):
```
Comment (taleinat) [inline Lib/xml/dom/pulldom.py:305]: Is this addition necessary? Does it do anything?

File: Lib/test/test_pulldom.py (hunk 7 above)
Comment (matrixise) [reply Lib/test/test_pulldom.py:121]: Don’t change the code when it’s no my a requirement. Keep it as simple as possible. Thanks

**Comment by jfgossage** [comment]
What is meant by the phrase "no my a requirement"? Can you point me at
documentation that explains the concept?

File: Lib/test/test_pulldom.py (hunk 7 above)
Comment (scoder) [reply Lib/test/test_pulldom.py:121]: This probably meant "Don't change the code when it's not a requirement", i.e. avoid unnecessary changes and try to keep them minimal for what you want to achieve.

**Comment by csabella** [comment]
@jgossage, this is still tagged as 'Awaiting Changes'.  If not done yet, please address the code reviews and let @taleinat know when it's ready for review.  Thanks!

**Comment by csabella** [comment]
I'm going to close this pull request as abandoned.  The merge is from an unknown repository and there hasn't been any response to the code review in over a year.  Please reopen if it shouldn't be closed.  Thanks!
